### DISCUSSION TITLE
Streamline the process of creating a Malachite-based application

### DISCUSSION THREAD

**Discussion by romac** [circlefin/malachite#549] [Ideas]
This proposal outlines a strategy to streamline the process of defining and running applications on top of Malachite.

## Background

Currently, implementing a Malachite application requires multiple manual steps and detailed knowledge of the framework's internals.
This creates unnecessary complexity and potential barriers to adoption.

## Current implementation

To develop an application on Malachite, the following steps are currently required:

1. Implement `malachite_common::Context` and associated traits
2. Implement `malachite_actors::util::codec::NetworkCodec`
3. Implement `Host` actor that can receive `malachite_actors::host::HostMsg<_>` messages
4. Implement `malachite_node::Node`
   1. Initialize metrics
   2. Spawn a Mempool actor (optional)
   3. Spawn a GossipConsensus actor
   4. Spawn a Host actor
   5. Spawn a BlockSync actor
   6. Spawn a Consensus actor
   7. Spawn a Node actor

## Proposed solution

The solution introduces three distinct integration patterns (actor, channel, and gRPC) through four new crates:

### Common functionality (`malachite-app`)

1. Create a new `malachite-app` crate providing core functionality:
   - Re-export `malachite-common` under `common` module
   - Export `NetworkCodec` trait
2. Rename `malachite_node::Node` trait to `Application` and move it to the `malachite-app` crate

### Actor-based apps (`malachite-app-actor`)

1. Create a new `malachite-app-actor` crate
   - Exports `malachite_actors` crate under `actors` module
2. Centralize actor spawn functions for spawning the actors listed above (except the host)
3. Provide a `run` function in `malachite-app-actor` which:
   - Takes the `Context` and `NetworkCodec`
   - Takes a closure for spawning the Host actor
   - Spawns all actors and return a reference to the Node actor

### Channel-based apps (`malachite-app-channel`)

1. Create a new `malachite-app-channel` crate
2. Define an internal `Host` actor which forwards messages through channel(s)
3. Provide a `run` function which returns the channel(s) for the app to make use of

### gRPC-based apps (`malachite-app-grpc`)

1. Create a new `malachite-app-grpc` crate
2. Define a `Host` actor which performs gRPC calls to a server
3. Provide a `run` function which takes the gRPC server address and starts the host

## Benefits

- Simplified application development
- Multiple integration patterns
- Reduced boilerplate code
- Clearer separation of concerns
- Improved maintainability

## Open Questions

0. How does this impact https://github.com/informalsystems/malachite/issues/373, and vice-versa?
2. Should the `Host` actor be renamed to `Application`?
   - Might be confusing if we have an `Application` trait as well.
   - Should the `Application` trait stay named Node so that we can use `Application` for what is now the host?
3. Should the `NetworkCodec` trait be moved to `malachite-common`?
   - It is only ever used in the `malachite-actors` crate
5. Is it possible to implement both the `Context` and `Application` traits on the same struct? Or is it more flexible to define two structures, one for the context and one for the app?
   - Probably best to keep them separate
6. Should we also propose a trait-based interface, in addition to the actor, channel and RPC interfaces?

## Related concerns

- For users wanting to write their own node based on `malachite-consensus`, we should ensure that the crate re-exports everything needed, without users having to pull multiple crates.

## Next Steps

1. [ ] Review and address open questions
3. [ ] Create detailed technical specifications
8. [ ] Implement proof-of-concept
9. [ ] Gather feedback

**Comment by greg-szabo** [comment]
I don't have the full picture yet, so I'm going to leave notes here before I jump to conclusions/PRs.

Generic thoughts:
I'm trying to describe the traits we have so I have an easier time remembering them. We use very overloaded terms, so I'm going to go overboard renaming them to spur ideas.
* `Context` - it's so generic that I question myself each time I come across it. Application context? Network context? Based on its doc comment, it "describes consensus datatypes for the consensus engine". So, maybe, the name should reflect that. `ConsensusData`, `CnsData`, etc. Even `ConsensusContext` would be helpful.
* `Host` - as far as I understand, Host is not a full-blown trait in itself. (I'm looking at `HostRef<Ctx>`). It refers to an Actor (as defined by ractor) as something that can handle HostMsg types. As such, it is most valuable in Actor-based implementations (`malachite-actor-app`) and maybe it should be moved closer to that. But this would only make sense if we had a Host trait that is unrelated to the ractor framework, so a channel-based app (`malachite-app-channel`) could implement it without knowing anything about ractor. This is not the case and for the short term we should accept that ractor will be needed in both Actor-based and channel-based applications. In which case, it could stay in actors as is.

Side note: I feel that we should strive to use ractor internally but provide interfaces (crates, structs, etc) that don't depend on them so the user can choose to use a different framework. I'm not sure how feasible that is, but it's a good stretch goal.

Back to the Host trait: If the topic is the Host trait defined in `malachite_starknet_host::host::Host`, then I got confused and you can ignore the above. Maybe we should put a comment on top of that trait. I believe that is defined by Starknet and we can't change it, right?

* If I want to describe the `Host` trait, I would say something along the line of: it is the client (or the "other party") for the Consensus Engine. The interface the user application can use to interact with the Consensus Engine. Unfortunately, most examples that come to mind are also overloaded. `ConsensusInterface`, `ConsensusClient`, maybe `ConsensusConnector`? If it was called `ConsensusClient`, I guess the engine is supposed to be called `ConsensusServer`. But it's a misnomer, as client-server interactions are one-way in the classic sense. (Interestingly, I ran into this regarding ABCI++, which fits into the client-server framework.) I'm not sure we want to rely on that. I like `ConsensusConnector` the most because it hints at what it is used for. You, the user, can connect to the Consensus Engine with it.

* `Node` trait: I agree that it should be renamed `Application` or `App` trait. It encompasses all parts of the application, and I was looking to implement that for the Actor-based implementation. 

* I think I have some reservations about the channel-based app implementation. Does that use the Host trait but not other parts of the malachite framework? (No Actors?) As I mentioned above, I believe we need actors for the host trait right now. Please clarify. Maybe I don't understand something.

Regarding the open questions:

* #373 concerns renaming crates, which is yet another refactoring. Based on the list of changes already outlined, I think we will have multiple PRs refactoring the codebase. I would put them in a convenient place on this same list, maybe at the end.

* Yup, `NodeCodec` is better in `malachite-common`. I wish there were a more generic trait defined in the global Rust ecosystem we could use so we give seamless integration to some crates. I checked some other crates that do encoding and decoding. Apparently, everyone does their own thing. Some separate it into an `Encoding` and `Decoding` trait. I think we're good as we are.

* My current opinion about renaming: `Node` -> `Application`; `Host` -> `ConsensusConnector`; `Context` -> `ConsensusData`.

* Regarding `Application+Context` struct: I think it's overkill to implement both traits on the same struct.

* Trait-based interface: First, I thought the `Node` (`Application`) does that, before understanding the underlying actor dependencies. Now, I'm not sure if it's possible without ractor. I'll have more thoughts about it as we implement the different models.

**Reply by romac** [reply]
> * `Context` - it's so generic that I question myself each time I come across it. Application context? Network context? Based on its doc comment, it "describes consensus datatypes for the consensus engine". So, maybe, the name should reflect that. `ConsensusData`, `CnsData`, etc. Even `ConsensusContext` would be helpful.

Agreed, it's a bit generic but it's short which is useful since it's present in pretty much every type bound ever in the codebase. expose it under a more explicit name from the `malachite-app` crate, perhaps `ConsensusContext` or `ConsensusSpecs`?

> * `Host` - as far as I understand, Host is not a full-blown trait in itself. (I'm looking at `HostRef<Ctx>`). It refers to an Actor (as defined by ractor) as something that can handle HostMsg types. As such, it is most valuable in Actor-based implementations (`malachite-actor-app`) and maybe it should be moved closer to that. But this would only make sense if we had a Host trait that is unrelated to the ractor framework, so a channel-based app (`malachite-app-channel`) could implement it without knowing anything about ractor. This is not the case and for the short term we should accept that ractor will be needed in both Actor-based and channel-based applications. In which case, it could stay in actors as is.

Yeah host is the most overloaded term here. Right now it denotes three things:
- In the abstract, the host is the component in the software system implementing a Malachite node that interfaces with consensus, handles its requests (eg. for assembling a value to propose), and keeps track of various things (blocks, streamed proposal parts, etc). I think "application" is a better term for this, being more commonly seen than "host", eg. in Comet.
- In the code, we have two instances of it:
  - 1. In the engine itself (`malachite-actors`), it is represented by the `HostMsg` enum which in the abstract describes all requests that abstract host must be able to fulfill. It also describes exactly the messages that the actor actually implementing the host must be able to handle. The `HostRef` type is an alias for an `ActorRef` through which consensus can send `HostMsg` to the actor.
  - 2. In the Starknet host definition (`malachite-starknet-host`), it is a concrete actor that can handle `HostMsg` messages and from which we can get a `HostRef` for consensus to send messages through.
  - 3. In that same crate, it also describes the interface between the "host" in the Malachite sense and the Starknet sequencer. That is very confusing and we should definitely rename it to something else (eg. `Sequencer` or something more explicit like this). For the purpose of this proposal, we shall ignore it as it is a Starknet-specific concern.

> Side note: I feel that we should strive to use ractor internally but provide interfaces (crates, structs, etc) that don't depend on them so the user can choose to use a different framework. I'm not sure how feasible that is, but it's a good stretch goal.

This is exactly the point of exposing a channel-based interface, which hides the whole actor framework behind regular Tokio channels. Users not wanting to implement a full blown actor would instead be able to depend on the `malachite-app-channel` crate which would provide a channel-based interface through which the host/application can receive and reply to `HostMsg`s send from consensus without ever seeing mention of `raptor` or any related terminology.

> Back to the Host trait: If the topic is the Host trait defined in `malachite_starknet_host::host::Host`, then I got confused and you can ignore the above. Maybe we should put a comment on top of that trait. I believe that is defined by Starknet and we can't change it, right?

In the context of this proposal, the host is related the Malachite <-> application interface and not to the Starknet sequencer interface. The latter is indeed defined by Starknet and not under consideration in this proposal.
 
> * If I want to describe the `Host` trait, I would say something along the line of: it is the client (or the "other party") for the Consensus Engine. The interface the user application can use to interact with the Consensus Engine. 

That's a very good definition!

> Unfortunately, most examples that come to mind are also overloaded. `ConsensusInterface`, `ConsensusClient`, maybe `ConsensusConnector`? If it was called `ConsensusClient`, I guess the engine is supposed to be called `ConsensusServer`. But it's a misnomer, as client-server interactions are one-way in the classic sense. (Interestingly, I ran into this regarding ABCI++, which fits into the client-server framework.) I'm not sure we want to rely on that. I like `ConsensusConnector` the most because it hints at what it is used for. You, the user, can connect to the Consensus Engine with it.

That's a good observation. I feel like the client-server paradigm applies here but in reverse, where consensus is the client and the application is the server. That said, I too am not particularly fond of this nomenclature, and I like your "connector" idea better!

> * `Node` trait: I agree that it should be renamed `Application` or `App` trait. It encompasses all parts of the application, and I was looking to implement that for the Actor-based implementation.

Glad we agree on that :)

> * I think I have some reservations about the channel-based app implementation. Does that use the Host trait but not other parts of the malachite framework? (No Actors?) As I mentioned above, I believe we need actors for the host trait right now. Please clarify. Maybe I don't understand something.

No it would not use the `malachite_starknet_host::Host` trait, that one is Starknet-specific.
What it would do instead is provide the application with a channel through which the application can receive requests from consensus (what we currently call `HostMsg`) and fulfill those. If a reply is needed, the request will contain a oneshot Tokio channel through which it can send a reply back to consensus. It would therefore not need to expose any of the actor-based internals.

> Regarding the open questions:
> 
> * [code: Restructure the repository and crates #373](https://github.com/informalsystems/malachite/issues/373) concerns renaming crates, which is yet another refactoring. Based on the list of changes already outlined, I think we will have multiple PRs refactoring the codebase. I would put them in a convenient place on this same list, maybe at the end.

Just leaving a note here that will need to freeze the codebase for a few days so that I can proceed with the renaming without creating humongous merge conflicts.

> * Yup, `NodeCodec` is better in `malachite-common`. I wish there were a more generic trait defined in the global Rust ecosystem we could use so we give seamless integration to some crates. I checked some other crates that do encoding and decoding. Apparently, everyone does their own thing. Some separate it into an `Encoding` and `Decoding` trait. I think we're good as we are.

Alternatively, we could introduce a `malachite-network` crate which expose that trait and other network-related utilities. I need to think about what else we could put there.

> * My current opinion about renaming: `Node` -> `Application`; `Host` -> `ConsensusConnector`; `Context` -> `ConsensusData`.

- `Node` -> `Application`: Agreed
- `Host` -> `ConsensusConnector`: Agreed
- `Context` -> `ConsensusData`: I would rather leave it as `Context` internally but expose it under `ConsensusContext` from the `malachite-app` crate. What do you think?

> * Regarding `Application+Context` struct: I think it's overkill to implement both traits on the same struct.

Agreed.

> * Trait-based interface: First, I thought the `Node` (`Application`) does that, before understanding the underlying actor dependencies. Now, I'm not sure if it's possible without ractor. I'll have more thoughts about it as we implement the different models.

Much like with the channel-based interface, it should be possible to expose a trait-based interface, where users implement a single trait, with a method corresponding to each message to be handled. This would also hide all the actor-based internals, much like with the channel based interface. The tradeoff between channel-based and trait-based, is that the former gives you more flexibility for processing messages, by allowing the user to write their own event loop and deal with messages however they like (potentially handling them concurrently or queuing them somewhere), whereas with a trait-based implementation, you typically relinquish that flexibility. I think this type of interface is not strictly needed and can be done at a later stage, as you suggest.

**Comment by romac** [comment]
One more thing we could do to streamline this process further is to provide a default implementation of the `Vote` and `Proposal` types for applications that do not need custom versions of those. Since these two datatypes are mostly implementation details of Tendermint, I do not expect that an application would to customize those very often. We could still make it possible to override these defaults if needed.

Here is how it might work:

Because default type members are still unstable, we cannot do the obvious:

```rust
use core::marker::PhantomData;

struct SignedMessage<Msg>(PhantomData<Msg>);

enum DefaultVote {
    Prevote
}

struct DefaultProposal {}

trait Context {
  type Vote = DefaultVote;
  type Proposal = DefaultVote;
  type Value;
  
  fn new_prevote(&self) -> Self::Vote {
    DefaultVote::Prevote
  }

  fn sign_vote(&self, vote: Self::Vote) -> SignedMessage<Self::Vote>;
  
  // etc.
}
```

Instead we might be able to do something like this, at the expense of a bunch of boilerplate:

```rust
use core::marker::PhantomData;

struct SignedMessage<Msg>(PhantomData<Msg>);

enum DefaultVote {
    Prevote
}

struct DefaultProposal {}

trait Context {
  type Vote;
  type Proposal;
  type Value;
  
  fn new_prevote(&self) -> Self::Vote;
  fn sign_vote(&self, vote: Self::Vote) -> SignedMessage<Self::Vote>;
  
  // etc.
}

trait DefaultContext {
  type Value;
  
  fn sign_vote(&self, vote: DefaultVote) -> SignedMessage<DefaultVote>;
  
  // etc.
}

impl<T> Context for T where T: DefaultContext {
    type Vote = DefaultVote;
    type Proposal = DefaultProposal;
    
    type Value = T::Value;
    
    fn new_prevote(&self) -> DefaultVote {
        DefaultVote::Prevote
    }
    
    fn sign_vote(&self, vote: DefaultVote) -> SignedMessage<DefaultVote> {
        <T as DefaultContext>::sign_vote(self, vote)
    }
}
```

Applications would then only need to implement the `DefaultContext` trait instead of `Context`.

---

One open question is whether or not to also provide a default Protobuf-based implementation of `NetworkCodec` for these two built-in types?

**Reply by greg-szabo** [reply]
Whoah, that's some excellent stuff with type message defaults. Since it mostly touches the `malachite-common` crate and we only need to make a few changes there, we could do it right after the already agreed changes. I consider it a new feature, which means there are no real dependents right now.

I've created a list of proposed steps that provides more details and tries to show the dependencies of each task. You can see it [here](https://docs.google.com/document/d/1_N-Csenj8RgTFUsi_zPjHJppFULoVuMw8kJiNH5necI/edit). (I can move it to GitHub, if needed, but it was cleaner to work in a doc.) I've integrated all that we discussed so far.

Legend:
* green highlight: new work that we can immediately work on
* yellow highlight: refactoring: easy work, but it requires a quick turnaround in review or a code freeze so it doesn't mess up everyone else's work
* no highlight: more complex work, that usually integrates new things with existing libraries. (Dependencies have no highlights.)

In the doc, tasks after each other usually depend on each other. (Except if they are highlighted green.)

Some new proposals coming from the doc:
* Should we rename `HostMsg` to `ConnectorMsg`? Since the idea of "Host" is renamed to "Connector", it might reduce confusion. I'm open to ideas. (`ConsensusMsg` is taken.)
* The Host/Connector for the channel-based and gRPC-based crates is still a bit hazy for me. I understand that the crates' `run` function will return Tokio channels or some kind of IO handle for gRPC, but I'm not sure yet how to define a "Host/Connector" for them. I think it will become more evident as I look into `spawn_host_actor` a bit more.

Edit: the tasks in the doc can be easily correlated to one task = one PR.

**Comment by andynog** [comment]
Hi guys, I was just going through all the discussion. I don't have much too add on top of what that has been discussed. Just a few comments:

- On renaming, I like the idea of calling `ConsensusContext` and the other ones too.
```
Node -> Application: Agreed
Host -> ConsensusConnector: Agreed
Context -> ConsensusData: I would rather leave it as Context internally but expose it under ConsensusContext from the malachite-app crate. What do you think?
```

- Related to the `client/server` paradigm comment
```
 I feel like the client-server paradigm applies here but in reverse, where consensus is the client and the application is the server. That said, I too am not particularly fond of this nomenclature, and I like your "connector" idea better!```
```
 I agree, we should avoid renaming things using client or server names. Using an actor model it is more related to messaging paradigms if that's what we want. The question should be, if we try to force this into a client/server paradigm will it add constraints on future uses cases that could open up if we stick with the actor model paradigm ? If so, we should consider this fact on renaming things and structuring. 

I couldn't access Greg's doc, but what I could help in the future is to go over the refactoring and try to implement an app and provide feedback if there are still pain points or not so clear or brittle design patterns.

**Reply by romac** [reply]
> I couldn't access Greg's doc, but what I could help in the future is to go over the refactoring and try to implement an app and provide feedback if there are still pain points or not so clear or brittle design patterns.

Yeah that would be really helpful, thanks :)

**Reply by greg-szabo** [reply]
I gave you access when you sent the request e-mail, please check again.

Edit: the disadvantages of using Google Docs... :(
